marlowe-chain-sync/app/Main.hs

```diff
@@ -1,86 +1,58 @@
-    runNodeClientSupervised = withAsync runNodeClient' \a -> do
-      result <- waitCatch a
-      case result of
-        Right _ -> pure ()
-        Left exception -> do
-          connectionAttempts <- atomically do
-            connectionAttempts <- readTVar tConnectionAttempts
-            writeTVar tConnectionAttempts $ connectionAttempts + 1
-            pure connectionAttempts
-          let
-            timeoutSeconds :: Int
-            timeoutSeconds = min 15 $ floor @Double $ exp $ fromIntegral connectionAttempts
-          let
-            msg = if connectionAttempts == 0
-              then "Connection to local node lost."
-              else "Connection attempt #" <> show connectionAttempts <> " failed."
-          hPutStrLn stderr $ msg <> " Attempting to reconnect in " <> show timeoutSeconds <> "s."
-          hPutStrLn stderr $ "Message: " <> displayException exception
-          threadDelay $ 1_000_000 * timeoutSeconds
-          runNodeClientSupervised
+run :: Options -> IO ()
+run Options{..} = do
+  connectionResult <- Connection.acquire (fromString databaseUri)
+  connection <- either
+    (fail . maybe "Failed to connect to database" (unpack . decodeUtf8))
+    pure
+    connectionResult
+  genesisConfigResult <- runExceptT do
+    hash <- ExceptT $ pure $ decodeAbstractHash genesisConfigHash
+    (hash,) <$> withExceptT
+      (const "failed to read byron genesis file")
+      (Byron.mkConfigFromFile (toByronRequiresNetworkMagic networkId) genesisConfigFile hash)
+  (hash, genesisConfig) <- either (fail . unpack) pure genesisConfigResult
+  chainSync <- atomically $ mkChainSync ChainSyncDependencies
+    { localNodeConnectInfo
+    , databaseQueries = hoistDatabaseQueries
+        (either throwQueryError pure <=< flip Session.run connection)
+        PostgreSQL.databaseQueries
+    , persistRateLimit
+    , genesisConfigHash = abstractHashToBytes hash
+    , genesisConfig
+    }
+  runChainSync chainSync
+  where
+    throwQueryError (Session.QueryError _ _ err) = error $ show err
 
-    runNodeClient' :: IO ()
-    runNodeClient' =
-      runNodeClient connectionInfo getHeaderAtPoint getIntersectionPoints \event -> atomically do
-        writeTVar tConnectionAttempts 0
-        case event of
-          RollForward (BlockInMode (Block (BlockHeader slotNo hash _) _) _) tip -> do
-            writeTVar tpoint $ ChainPoint slotNo hash
-            writeTVar ttip tip
-          RollBackward point tip                                                -> do
-            writeTVar tpoint point
-            writeTVar ttip tip
-      where
-        connectionInfo = LocalNodeConnectInfo
-          { localConsensusModeParams = CardanoModeParams $ EpochSlots 21600
-          , localNodeNetworkId = networkId
-          , localNodeSocketPath = nodeSocket
-          }
-        getHeaderAtPoint _ = pure Origin
-        getIntersectionPoints _ _ = pure []
+    localNodeConnectInfo :: LocalNodeConnectInfo CardanoMode
```

Comment:
_General comment (FYI only):_ Since we may run this eventually with Hydra, where the node connection has a different type, we may want to keep the node-specific functions as separate as possible from the pure computations that use them.